cmd/kubectl-gadget/main.go

```diff
@@ -87,21 +95,41 @@ func main() {
 			fmt.Fprintf(os.Stderr, "Error: %v\n", err)
 			os.Exit(1)
 		}
-		runtime.InitDeployInfo()
+		grpcRuntime.InitDeployInfo()
 	}
 
 	namespace, _ := utils.GetNamespace()
-	runtime.SetDefaultValue(gadgets.K8SNamespace, namespace)
-
-	hiddenColumnTags := []string{"runtime"}
-	common.AddCommandsFromRegistry(rootCmd, runtime, hiddenColumnTags)
+	grpcRuntime.SetDefaultValue(gadgets.K8SNamespace, namespace)
 
 	// Advise category is still being handled by CRs for now
 	rootCmd.AddCommand(advise.NewAdviseCmd())
 
-	rootCmd.AddCommand(common.NewSyncCommand(runtime))
+	rootCmd.AddCommand(common.NewSyncCommand(grpcRuntime))
+
+	// Deploy and update our local registry before we add commands from it
+	if deploy {
+		err := runDeploy(os.Stderr)
+		if err != nil {
+			log.Errorf("deploying Inspektor Gadget: %v\n", err)
+			os.Exit(1)
+		}
+		grpcRuntime.InitDeployInfo()
+	}
+
+	hiddenColumnTags := []string{"runtime"}
+	common.AddCommandsFromRegistry(rootCmd, grpcRuntime, hiddenColumnTags)
+
+	err := rootCmd.Execute()
+
+	if deploy {
+		err := runUndeploy(os.Stderr)
+		if err != nil {
+			log.Errorf("undeploying Inspektor Gadget: %v\n", err)
+			os.Exit(1)
+		}
+	}
```

Comment:
The output still behaves like before: There is no output of the `err` returned by `rootCmd.Execute`. But I could add that

Problem with `defer` is that it will not run when we are exiting with `os.Exit()`. See Line 133:
```golang
	if err != nil {
		os.Exit(1)
	}
```